[mlir] share argument attributes interface between calls and callables

jeanPerier opened the pull request:
First patch of this [RFC](https://discourse.llvm.org/t/mlir-rfc-adding-argument-and-result-attributes-to-llvm-call/84107) to add argument and result attributes on llvm.call (and llvm.invoke).

This patch shares core interface methods dealing with argument and result attributes from CallableOpInterface with the CallOpInterface and makes them mandatory to gives more consistent guarantees about concrete operations using these interfaces. 

Add optional `arg_attrs` and `res_attrs` attributes to operations using these interfaces that did not have that already.
They can then re-use the common "rich function signature" printing/parsing helpers if they want (for the LLVM dialect, this is done in the next patch).

[edit: change the inheritance level after RFC feedback]
[edit 2: remove extra interface and use listconcat instead to share methods. Make them mandatory].

jeanPerier committed:
[mlir] share argument attributes interface between calls and callables

jeanPerier committed:
rename and change inheritance level

River707 commented on mlir/include/mlir/Interfaces/CallImplementation.h:
```diff
@@ -0,0 +1,88 @@
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef MLIR_INTERFACES_CALLIMPLEMENTATION_H
```
Can you just merge this file into CallInterfaces.h? I know FunctionImplementation,h is its own file, but that's honestly a bit of historical fact at this point.

River707 commented on mlir/include/mlir/Interfaces/CallInterfaces.td:
```diff
@@ -17,6 +17,53 @@
+
+/// Interface for operations with arguments attributes (both call-like
+/// and callable operations).
+def ArgumentAttributesOpInterface : OpInterface<"ArgumentAttributesOpInterface"> {
```
I mentioned this in the RFC, but I'd avoid the additional interface all together. I don't think there is a lot of benefit added here over the cost on the generated code. Can this just be a list of additional methods that get added to the interface methods in CallOpInterface and CallableOpInterface?

River707 commented on mlir/lib/Interfaces/CallImplementation.cpp:
```diff
@@ -0,0 +1,178 @@
+//===- CallImplementation.pp - Call and Callable Op utilities -------------===//
```
Can you merge this file into CallInterfaces.cpp?

River707 reviewed (changes requested):
The general code movement seems good!

jeanPerier committed:
PR comments: remove new interface and added files

Share methods via listconcat instead.
Make methods mandatory (no default impl).
Add arg_attrs and res_attrs to all concrete operations inheriting from
CallOpInterface and CallableOpInterface.

jeanPerier commented:
Thanks for the feedback @River707!

I updated this PR based on your comments and the RFC to make the methods mandatory in both CallOpInterafce and CallableOpInterface without using an extra OpInterface. This patch is now adding new optional `arg_attrs` and `res_attrs` attributes to all concrete operations inheriting these interfaces.

jeanPerier committed:
update toy dialect too

gysit reviewed (approved):
LGTM from my end.

Thanks for working on this feature! I added some nits. Please pick the ones that make sense.

jeanPerier committed:
Apply suggestions from code review

Co-authored-by: Tobias Gysi <tobias.gysi@nextsilicon.com>

jeanPerier committed:
rename parseFunctionSignature to parseFunctionSignatureWithArguments

jeanPerier commented:
> LGTM from my end.
> 
> Thanks for working on this feature! I added some nits. Please pick the ones that make sense.

Thanks for the review! The nits made sense to me. I applied them.

River707 commented on mlir/include/mlir/Interfaces/CallInterfaces.h:
```diff
@@ -14,6 +14,7 @@
 #ifndef MLIR_INTERFACES_CALLINTERFACES_H
 #define MLIR_INTERFACES_CALLINTERFACES_H
 
+#include "mlir/IR/OpImplementation.h"
```
OpImplementation.h pulls in a lot of extra junk, can you use forward declares here instead for whatever is needed?

jeanPerier replied:
No, because of `OpAsmParser::Argument`  usage, I think it is not possible to forward declare nested types.

Either that class should be hoisted in a new OpAsmParserArgument class (a bit outside of the scope of this patch), or the helpers need to be in a different header (maybe that is why FunctionImplementation.h exists).

What do you think, is it worth splitting the header to avoid that include?

jeanPerier replied:
I plan to not split the header. Will likely merge on Monday, please tell me if you think splitting the header is better.

River707 commented on mlir/include/mlir/Interfaces/CallInterfaces.td:
```diff
@@ -31,7 +73,7 @@ def CallOpInterface : OpInterface<"CallOpInterface"> {
   let cppNamespace = "::mlir";
 
-  let methods = [
+  let methods = !listconcat(ArgumentAttributesMethods.methods, [
```
You should be able to just use the operator for this `] # ArgumentAttributesMethods.methods`

River707 commented on mlir/include/mlir/Interfaces/CallInterfaces.td:
```diff
@@ -96,7 +138,7 @@ def CallableOpInterface : OpInterface<"CallableOpInterface"> {
   let cppNamespace = "::mlir";
 
-  let methods = [
+  let methods = !listconcat(ArgumentAttributesMethods.methods, [
```
Same here.

jeanPerier committed:
use concat operator instead of lisconcat

jeanPerier commented:
> Silly question: will this allow arg/result attributes on intrinsics?

Not through this interface change because llvm.call_intrinsic does not use the CallOpInterface.

So if this is desired, this should be added separately on the llvm.call_intrinsic.
